File: src/commands/clear.ts
import { Command } from 'commander';
import { getContext, handleError } from './shared.js';
import { getAgent, deleteClaimsByAgent, deleteClaim, createMessage, getClaimsByAgent } from '../db/queries.js';
import { resolveAgentRef } from '../core/context.js';

/**
 * Strip # prefix from bd/issue values.
 */
function stripHash(value: string): string {
  return value.startsWith('#') ? value.substring(1) : value;
}

export function clearCommand(): Command {
  return new Command('clear')
    .description('Clear claims for an agent')
    .argument('<agent>', 'agent name (e.g., @alice)')
    .option('--file <path>', 'clear a specific file claim')
    .option('--bd <id>', 'clear a specific beads issue claim')
    .option('--issue <id>', 'clear a specific GitHub issue claim')
    .action((agentArg: string, options, cmd) => {
      try {
        const { db, jsonMode, projectConfig } = getContext(cmd);
        const agentId = resolveAgentRef(agentArg, projectConfig ?? null);

        // Verify agent exists
        const agent = getAgent(db, agentId);
        if (!agent) {
          throw new Error(`Agent not found: @${agentId}`);
        }

        let cleared = 0;
        const clearedItems: string[] = [];

        // Clear specific claims if flags provided
        if (options.file) {
          if (deleteClaim(db, 'file', options.file)) {
            cleared++;
            clearedItems.push(options.file);
          }
        }

        if (options.bd) {
          const pattern = stripHash(options.bd);
          if (deleteClaim(db, 'bd', pattern)) {
            cleared++;
            clearedItems.push(`bd:${pattern}`);
          }
        }

        if (options.issue) {
          const pattern = stripHash(options.issue);
          if (deleteClaim(db, 'issue', pattern)) {
            cleared++;
            clearedItems.push(`issue:${pattern}`);
          }
        }

        // If no specific flags, clear all claims for the agent
        if (!options.file && !options.bd && !options.issue) {
          const existingClaims = getClaimsByAgent(db, agentId);
          cleared = deleteClaimsByAgent(db, agentId);
          for (const claim of existingClaims) {
            if (claim.claim_type === 'file') {
              clearedItems.push(claim.pattern);
            } else {
              clearedItems.push(`${claim.claim_type}:${claim.pattern}`);
            }
          }
        }

        // Post message if anything was cleared
        if (cleared > 0) {
          createMessage(db, {
            from_agent: agentId,
            body: `cleared claims: ${clearedItems.join(', ')}`,
            mentions: [],
          });
        }

        if (jsonMode) {
          console.log(JSON.stringify({
            agent_id: agentId,
            cleared,
            items: clearedItems,
          }));
        } else {
          if (cleared === 0) {
            console.log(`No claims to clear for @${agentId}`);
          } else {
            console.log(`@${agentId} cleared ${cleared} claim${cleared !== 1 ? 's' : ''}:`);
            for (const item of clearedItems) {
              console.log(`  ${item}`);
            }
          }
        }

        db.close();
      } catch (error) {
        handleError(error);
      }
    });
}

File: internal/command/clear.go
package command

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/adamavenir/fray/internal/db"
	"github.com/adamavenir/fray/internal/types"
	"github.com/spf13/cobra"
)

// NewClearCmd creates the clear command.
func NewClearCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "clear <agent>",
		Short: "Clear claims for an agent",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			ctx, err := GetContext(cmd)
			if err != nil {
				return writeCommandError(cmd, err)
			}
			defer ctx.DB.Close()

			agentID, err := resolveAgentRef(ctx, args[0])
			if err != nil {
				return writeCommandError(cmd, err)
			}
			agent, err := db.GetAgent(ctx.DB, agentID)
			if err != nil {
				return writeCommandError(cmd, err)
			}
			if agent == nil {
				return writeCommandError(cmd, fmt.Errorf("agent not found: @%s", agentID))
			}

			file, _ := cmd.Flags().GetString("file")
			bd, _ := cmd.Flags().GetString("bd")
			issue, _ := cmd.Flags().GetString("issue")

			cleared := int64(0)
			clearedItems := []string{}

			if file != "" {
				deleted, err := db.DeleteClaim(ctx.DB, types.ClaimTypeFile, file)
				if err != nil {
					return writeCommandError(cmd, err)
				}
				if deleted {
					cleared++
					clearedItems = append(clearedItems, file)
				}
			}
			if bd != "" {
				pattern := stripHash(bd)
				deleted, err := db.DeleteClaim(ctx.DB, types.ClaimTypeBD, pattern)
				if err != nil {
					return writeCommandError(cmd, err)
				}
				if deleted {
					cleared++
					clearedItems = append(clearedItems, fmt.Sprintf("bd:%s", pattern))
				}
			}
			if issue != "" {
				pattern := stripHash(issue)
				deleted, err := db.DeleteClaim(ctx.DB, types.ClaimTypeIssue, pattern)
				if err != nil {
					return writeCommandError(cmd, err)
				}
				if deleted {
					cleared++
					clearedItems = append(clearedItems, fmt.Sprintf("issue:%s", pattern))
				}
			}

			if file == "" && bd == "" && issue == "" {
				existing, err := db.GetClaimsByAgent(ctx.DB, agentID)
				if err != nil {
					return writeCommandError(cmd, err)
				}
				for _, claim := range existing {
					if claim.ClaimType == types.ClaimTypeFile {
						clearedItems = append(clearedItems, claim.Pattern)
					} else {
						clearedItems = append(clearedItems, fmt.Sprintf("%s:%s", claim.ClaimType, claim.Pattern))
					}
				}
				cleared, err = db.DeleteClaimsByAgent(ctx.DB, agentID)
				if err != nil {
					return writeCommandError(cmd, err)
				}
			}

			if cleared > 0 {
				body := fmt.Sprintf("cleared claims: %s", joinList(clearedItems))
				msg, err := db.CreateMessage(ctx.DB, types.Message{
					TS:        time.Now().Unix(),
					FromAgent: agentID,
					Body:      body,
					Mentions:  []string{},
				})
				if err != nil {
					return writeCommandError(cmd, err)
				}
				if err := db.AppendMessage(ctx.Project.DBPath, msg); err != nil {
					return writeCommandError(cmd, err)
				}
			}

			if ctx.JSONMode {
				payload := map[string]any{
					"agent_id": agentID,
					"cleared":  cleared,
					"items":    clearedItems,
				}
				return json.NewEncoder(cmd.OutOrStdout()).Encode(payload)
			}

			out := cmd.OutOrStdout()
			if cleared == 0 {
				fmt.Fprintf(out, "No claims to clear for @%s\n", agentID)
				return nil
			}
			plural := "s"
			if cleared == 1 {
				plural = ""
			}
			fmt.Fprintf(out, "@%s cleared %d claim%s:\n", agentID, cleared, plural)
			for _, item := range clearedItems {
				fmt.Fprintf(out, "  %s\n", item)
			}
			return nil
		},
	}

	cmd.Flags().String("file", "", "clear a specific file claim")
	cmd.Flags().String("bd", "", "clear a specific beads issue claim")
	cmd.Flags().String("issue", "", "clear a specific GitHub issue claim")
	return cmd
}
